lib/plenario_etl/worker.ex

```diff
@@ -1,290 +1,177 @@
+    load!(model, path, columns, constraints, fn pth ->
+      File.read!(pth)
       |> Poison.decode!()
     end)
   end
 
-  @doc """
-  """
-  def load_csv(meta, path, job, constraint) do
-    Logger.info("[#{inspect(self())}] [load_csv] Prep loader for csv at #{path}")
+  defp load_csv!(path, model, columns, constraints) do
+    Logger.info("using csv loader")
 
-    load_data(meta, path, job, constraint, fn path ->
-      File.stream!(path)
+    load!(model, path, columns, constraints, fn pth ->
+      File.stream!(pth)
       |> CSV.decode!(headers: true)
     end)
   end
 
-  @doc """
-  """
-  def load_tsv(meta, path, job, constraints) do
-    Logger.info("[#{inspect(self())}] [load_tsv] Prep loader for tsv at #{path}")
+  defp load_tsv!(path, model, columns, constraints) do
+    Logger.info("using tsv loader")
 
-    load_data(meta, path, job, constraints, fn path ->
+    load!(model, path, columns, constraints, fn pth ->
       File.stream!(path)
       |> CSV.decode!(headers: true, separator: ?\t)
     end)
   end
 
-  @doc """
-  Performs the work of loading a shapefile associated with a `Meta` instance.
-  """
-  def load_shape(meta, path, _job) do
-    Logger.info("[#{inspect(self())}] [load_shape] Unpacking shapefile at #{path}")
-    {:ok, file_paths} = :zip.unzip(String.to_charlist(path), cwd: '/tmp/')
-
-    Logger.info("[#{inspect(self())}] [load_shape] Looking for .shp file")
-
-    shp =
-      Enum.find(file_paths, fn path ->
-        String.ends_with?(to_string(path), ".shp")
-      end)
-      |> to_string()
-
-    Logger.info("[#{inspect(self())}] [load_shape] Prep loader for shapefile at #{shp}")
-    PlenarioEtl.Shapefile.load(shp, meta.name)
+  defp load_shp!(path, model, columns, constraints) do
+    {:erred, "not implemented"}
   end
 
-  defp load_data(meta, path, job, constraints, decode) do
-    Logger.info("[#{inspect self()}] [load_data] Chunking rows and spawning children")
-
-    decode.(path)
-    |> Stream.chunk_every(@chunk_size)
-    |> Enum.map(fn chunk ->
-      load_chunk!(meta, job, chunk, constraints) 
-    end)
-  end
+  defp load!(model, path, columns, constraints, decode) do
+    Logger.info("ripping file")
 
-  def async_load!(meta_id) do
-    meta = MetaActions.get(meta_id)
-    job = EtlJobActions.create!(meta)
-    # Set next import here
-    {:ok, job} = EtlJobActions.mark_started(job)
-
-    task = Task.async(fn ->
-      :poolboy.transaction(
-        :worker,
-        fn pid ->
-          try do
-            GenServer.call(pid, {:load, %{
-              meta_id: meta_id,
-              job_id: job.id
-            }}, :infinity)
-            EtlJobActions.mark_completed(job)
-          catch
-            :exit, message ->
-              Logger.error(inspect(message))
-              EtlJobActions.mark_erred(job, %{error_message: inspect(message)})
-          end
-        end,
-        :infinity
-      )
-    end)
+    results =
+      decode.(path)
+      |> Enum.reduce([], fn row, acc ->
```

Comment:
> I also cut out the parts that chunked the source document into their own processes -- it was a good idea but in reality it became a resource hog.

The chunking doesn't kick out any new processes anymore. It only restricts the amount of rows being worked on. I would add a similar windowing mechanism for this part of the code because from the looks of it there isn't any upper bounds to that accumulator.

Reply:
The reason for accumulating all the results is so I can switch between `succeeded`, `partial_success` and `erred`. On the last two, I'm dumping the error messages. If I didn't keep the tuples as is, then I would lose a lot of valuable information needed elsewhere.